ophyd/pseudopos.py

```diff
@@ -327,55 +642,169 @@ def target(self):
+        ValueError : on invalid positions
+        RuntimeError : if motion fails other than timing out
+        '''
         real_pos = self.forward(position)
 
+        # Clear all old statuses for not yet completed real motions
+        del self._real_waiting[:]
+
         # Remove the 'finished moving' callback, otherwise callbacks will
         # happen when individual motors finish moving
         moved_cb = kwargs.pop('moved_cb', None)
 
-        if self.sequential:
-            for real, value in zip(self._real, real_pos):
-                logger.debug('[sequential] Moving %s to %s (timeout=%s)',
-                             real.name, value, timeout)
-                if timeout <= 0:
-                    raise TimeoutError('Failed to move all positioners within '
-                                       '%s s'.format(timeout))
-
-                t0 = time.time()
-                try:
-                    real.move(value, wait=True, timeout=timeout, **kwargs)
-                except Exception:
-                    # TODO tag something onto exception message?
-                    raise
+        st = super().move(position, moved_cb=moved_cb, timeout=timeout)
 
-                elapsed = time.time() - t0
-                timeout -= elapsed
+        with self._finished_lock:
+            # ensure we don't get any motion complete messages before motion
+            # setup is finished
+            if self.sequential:
+                self._sequential_move(real_pos, timeout=timeout)
+            else:
+                self._concurrent_move(real_pos, timeout=timeout)
 
-        else:
-            del self._real_waiting[:]
-            self._real_waiting.extend(self._real)
+        if wait:
+            st.wait()
 
-            for real, value in zip(self._real, real_pos):
-                logger.debug('[concurrent] Moving %s to %s', real.name, value)
-                logger.debug('[concurrent] %s => %s', real.prefix, value)
-                real.move(value, wait=False, **kwargs)
+        return st
 
-        ret = Positioner.move(self, position, moved_cb=moved_cb, wait=wait,
-                              **kwargs)
+    @pseudo_position_argument
+    def forward(self, pseudo_pos):
+        '''Calculate a RealPosition from a given PseudoPosition
 
-        if self.sequential or (wait and not self.moving):
-            self._done_moving()
+        Must be defined on the subclass.
 
-        return ret
+        Parameters
+        ----------
+        pseudo_pos : PseudoPosition
+            The pseudo position input
 
-    def forward(self, pseudo_pos):
-        ''' '''
-        return self.RealPosition()
+        Returns
+        -------
+        real_position : RealPosition
+            The real position output
+        '''
+        # return self.RealPosition()
+        raise NotImplementedError()
 
-    def __call__(self, pseudo_pos):
-        return self.forward(pseudo_pos)
+    @pseudo_position_argument
+    def __call__(self, position):
```

Comment:
This surprised me a bit to find this.  The forward calculation is not _clearly_ what I would expect `p_pos(foo)` to do.